<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** Sean Nicolás González Vázquez y José Tonatiuh Navarro Silva.

**Fecha:** 10 de septiembre del 2020.

**Expediente** : 720083 y 723399
**Profesor:** Oscar David Jaramillo Zuluaga.

# Tarea 2: Clase Optimizacion de código

### El link donde se evidencia el trabajo es el siguiente: [Perfil GitHub Nicolas Gonzalez](https://github.com/nicoglez)

## Enunciado de tarea 
#### **Ejercicio 1**
1. Imagine una rutina contable utilizada en una librería. Funciona en una lista con sublistas, que se ven así:

Escriba un programa en Python, que devuelve una lista con tuplas de tamaño 2. Cada tupla consiste en el número de pedido y el costo total del pedido. El producto debe aumentarse en 10€ si el valor de la orden es inferior a 100,00 €. 
Escribe un programa Python usando unicamente las funciones **lambda y map**.

`orders = [ ["34587", "Learning Python, Mark Lutz", 4, 40.95], 
           ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]`

#### **Ejercicio 2**
2. La misma librería, pero esta vez trabajamos en una lista diferente. Las sublistas de nuestras listas se ven así: 
[número de orden, (número de artículo, cantidad, precio por unidad), ... (número de artículo, cantidad, precio por unidad)] 

`orders = [[1, ("5464", 4, 9.99), ("8274",18,12.99), ("9744", 9, 44.95)], 
	       [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
	       [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
           [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)] ]`

Escriba un programa que devuelva una lista de dos tuplas que tengan la información de (número de pedido, cantidad total de pedido). Utilice la función `Reduce`. Recuerde el precio del pedido del producto debe aumentarse en 10€ si el valor de la orden es inferior a 100,00 €.

#### **Ejercicio 3**
3. Resolver los anteriores ejercicios usando pandas (DataFrame)

#### **Ejercicio 4**
Use filter to eliminate all words that are shorter than 4 letters from a list of words
`list='the notion of a lambda function goes all the way back to the origin of computer science'`

#### **Ejercicio 5**
Use filter to determine the percentage of Fahrenheit temperatures in a list are within the range 32 to 80

#### **Ejercicio 6**
Use reduce to find the lower left corner (minimum x and minimum y value) for a list of point locations

### Resolucion de Ejercicios

In [8]:
# Importacion de librerias
# importacion de librerias
import pandas as pd
import numpy as np
from functools import reduce

### Ejercicio 1: 
1. Imagine una rutina contable utilizada en una librería. Funciona en una lista con sublistas, que se ven así:

Escriba un programa en Python, que devuelve una lista con tuplas de tamaño 2. Cada tupla consiste en el número de pedido y el costo total del pedido. El producto debe aumentarse en 10€ si el valor de la orden es inferior a 100,00 €. 
Escribe un programa Python usando unicamente las funciones **lambda y map**.

`orders = [ ["34587", "Learning Python, Mark Lutz", 4, 40.95], 
           ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]`

In [12]:
# Codigo Nicolas Glez
orders =  [["34587", "Learning Python, Mark Lutz", 4, 40.95], 
           ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]

conditional = lambda x: x+ 10 if (x < 100) else x # Funcion para aplicar la condicion
multi = lambda a,b : a*b # funcion para multiplicar

t = list(map(tuple, zip(*orders))) # Transponer tuplas
o = t[0], list(map(conditional,(tuple(map(multi, t[2], t[3]))))) #aplicar condicional y multiplicar y juntar tuplas
s1 = list(map(tuple, zip(*o))) # Regresar tuplas a posicion inicial

s1

[('34587', 163.8),
 ('98762', 284.0),
 ('77226', 108.85000000000001),
 ('88112', 84.97)]

In [ ]:
# Codigo Jose Tonatiuh

### **Ejercicio 2**
2. La misma librería, pero esta vez trabajamos en una lista diferente. Las sublistas de nuestras listas se ven así: 
[número de orden, (número de artículo, cantidad, precio por unidad), ... (número de artículo, cantidad, precio por unidad)] 

`orders = [[1, ("5464", 4, 9.99), ("8274",18,12.99), ("9744", 9, 44.95)], 
	       [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
	       [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
           [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)] ]`

Escriba un programa que devuelva una lista de dos tuplas que tengan la información de (número de pedido, cantidad total de pedido). Utilice la función `Reduce`. Recuerde el precio del pedido del producto debe aumentarse en 10€ si el valor de la orden es inferior a 100,00 €.

In [13]:
# Codigo Nicolas Glez
orders2 = [[1, ("5464", 4, 9.99), ("8274",18,12.99), ("9744", 9, 44.95)], 
          [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
          [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
          [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)] ]

conditional = lambda x: x+ 10 if (x < 100) else x # Funcion para aplicar la condicion
suma = lambda c,d : c + d # funcion para sumar

i = list(zip(*orders2))[0] # Guardar el indice
temp = list(map(lambda x: x[1:], orders2)) # Guardar las demas variables
p = list(map(lambda c: reduce(suma, list(map(conditional, list(map(lambda x: x[1]*x[2], c))))), temp)) # iterar sobre los datos

sol2 = list(zip(i, p)) # imprimir la solucion 
sol2

[(1, 688.3299999999999),
 (2, 504.46000000000004),
 (3, 374.79999999999995),
 (4, 502.57)]

In [ ]:
# Codigo Jose Tonatiuh

### **Ejercicio 3**
3. Resolver los anteriores ejercicios usando pandas (DataFrame)

In [14]:
# Solucion Ejercicio 3.1 con Pandas

# Codigo Nicolas Glez
orders3 =  [["34587", "Learning Python, Mark Lutz", 4, 40.95], 
           ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]

conditional = lambda x: x+ 10 if (x < 100) else x # Funcion para aplicar la condicion

orders3 = pd.DataFrame(orders3)
orders3 = orders3.drop(orders3.columns[[1]], axis = 'columns')
orders3['4'] = pd.DataFrame(map(conditional,orders3.iloc[:,1] * orders3.iloc[:,2]))
sol31 = orders3.drop(orders3.columns[[1,2]], axis = 'columns')
sol31


0       4
0  34587  163.80
1  98762  284.00
2  77226  108.85
3  88112   84.97

In [ ]:
# Codigo Jose Tonatiuh

In [17]:
# Solucion Ejercicio 3.2 con Pandas

# Codigo Nicolas Glez
orders4 = [[1, ("5464", 4, 9.99), ("8274",18,12.99), ("9744", 9, 44.95)], 
          [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
          [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
          [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)] ]

conditional = lambda x: x+ 10 if (x < 100 and x != 0) else x # Funcion para aplicar la condicion
multi = lambda a,b : a*b # funcion para multiplicar
suma = lambda c,d,e : c + d + e # funcion para sumar

orders4 = pd.DataFrame(orders4)
orders4.fillna(0, inplace = True)
temp2 = pd.DataFrame(orders4.iloc[:,1:])
t1, t2, t3 = (temp2.iloc[:, 0]), (temp2.iloc[:, 1]),(temp2.iloc[:, 2])

tx = pd.DataFrame({'A': [t1[:][0][1], t1[:][1][1], t1[:][2][1], t1[:][3][1]],
                   'B': [t1[:][0][2], t1[:][1][2], t1[:][2][2], t1[:][3][2]],
                   'C': [t2[:][0][1], t2[:][1][1], t2[:][2][1], t2[:][3][1]],
                   'D': [t2[:][0][2], t2[:][1][2], t2[:][2][2], t2[:][3][2]],
                   'E': [t3[:][0][1], t3[1], t3[2], t3[:][3][1]],
                   'F': [t3[:][0][2], t3[1], t3[2], t3[:][3][2]]})


tx2 = pd.DataFrame({'A': map(conditional, map(multi, tx['A'], tx['B'])), 
                    'B': map(conditional, map(multi, tx['C'], tx['D'])),
                    'C': map(conditional, map(multi, tx['E'], tx['F']))})

tx3 = pd.DataFrame({'Order Number': (orders4.iloc[:,0]),
                   'Total': (sum(tx2.loc[0]), sum(tx2.loc[1]), sum(tx2.loc[2]), sum(tx2.loc[3]))})
tx3

Order Number   Total
0             1  688.33
1             2  504.46
2             3  374.80
3             4  502.57

In [ ]:
# Codigo Jose Tonatiuh

### **Ejercicio 4**
Use filter to eliminate all words that are shorter than 4 letters from a list of words
`list='the notion of a lambda function goes all the way back to the origin of computer science'`

In [9]:
# Solucion Ejercicio 4

# Codigo Nicolas Glez
lista = 'the notion of a lambda function goes all the way back to the origin of computer science'
fw = lambda x: None if len(x)<4 else x

s4 = list(filter(fw, lista.split()))
s4

['notion',
 'lambda',
 'function',
 'goes',
 'back',
 'origin',
 'computer',
 'science']

In [ ]:
# Codigo Jose Tonatiuh

### **Ejercicio 5**
Use filter to determine the percentage of Fahrenheit temperatures in a list are within the range 32 to 80


In [10]:
# Solucion Ejercicio 5

# Codigo Nicolas Glez
np.random.seed(55555)
temperatures = np.random.uniform(25,110,30)
#print(temperatures)

fp = lambda x: x if x>=32 and x<=80 else None
        
s5 = list(filter(fp, temperatures))
s5

[35.97387961379455,
 63.211195349983264,
 73.90387696517755,
 36.982512376595906,
 68.75395211338406,
 45.86132066984196,
 36.58307493022292,
 56.322083650154724,
 33.9621930206693,
 45.29071317816652,
 74.86289938282341,
 66.51091382521807,
 46.831316244170544,
 44.55796498523391]

In [ ]:
# Codigo Jose Tonatiuh

### **Ejercicio 6**
Use reduce to find the lower left corner (minimum x and minimum y value) for a list of point locations

In [11]:
# Solucion Ejercicio 6

# Codigo Nicolas Glez
np.random.seed(55555)
point_xy = np.random.randint(0,100,[30,2])

f = np.minimum.reduce

s6 = list(reduce(lambda x,y : f([x,y]), point_xy))
s6

[0, 2]

In [ ]:
# Codigo Jose Tonatiuh

### Respuesta planteada en el ejercicio usando (markdown)


**Ejercicio 1** La respuesta del ejercicio 1 es, que al usar las herramientas de programacion pre expuestas, se encontro que la tupla que consiste del numero de pedido y el costo total del pedido que nos arrojo fue la siguiente.

```
[('34587', 163.8),
 ('98762', 284.0),
 ('77226', 108.85000000000001),
 ('88112', 84.97)]
 ```

**Ejercicio 2** La respuesta del ejercicio 2 es que se encontro que la tupla de dos elementos que consiste del numero de pedido y el costo total de mismo arrojada fue la siguiente.

```
[(1, 688.3299999999999),
 (2, 504.46000000000004),
 (3, 374.79999999999995),
 (4, 502.57)]
```

**Ejercicio 3.1** La respuesta del ejercicio 3.1 es que, al hacer el mismo ejercicio 1 con Pandas, se obtuvo el mismo resultado, pero con diferente procedimiento, teniendo la siguiente tabla como data frame:

|   	| 0     	| 4      	|
|---	|-------	|--------	|
| 0 	| 34587 	| 163.80 	|
| 1 	| 98762 	| 284.00 	|
| 2 	| 77226 	| 108.85 	|
| 3 	| 88112 	| 84.97  	|

**Ejercicio 3.2** La respuesta del ejercicio 3.2 es que, al hacer el mismo ejercicio 2 con Pandas, se obtuvo el mismo resultado, pero con diferente procedimiento, teniendo la siguiente tabla como data frame:

|   	| Order Number 	| Total  	|
|---	|--------------	|--------	|
| 0 	| 1            	| 688.33 	|
| 1 	| 2            	| 504.46 	|
| 2 	| 3            	| 374.80 	|
| 3 	|            4 	| 502.57 	|

**Ejercicio 4** La respuesta del ejercicio 4 es, que obtuvimos que las palabras en la lista comprendida que son menores a 4 letras son las siguientes.

```
['notion',
 'lambda',
 'function',
 'goes',
 'back',
 'origin',
 'computer',
 'science']
```

**Ejercicio 5** La respuesta del ejercicio 5 es, que al filtrar los datos que se tenian, encontramos que las temperaturas en grados Farenheit que se encuentran en el rango entre 32 y 80, son las siguientes.

```
[35.97387961379455,
 63.211195349983264,
 73.90387696517755,
 36.982512376595906,
 68.75395211338406,
 45.86132066984196,
 36.58307493022292,
 56.322083650154724,
 33.9621930206693,
 45.29071317816652,
 74.86289938282341,
 66.51091382521807,
 46.831316244170544,
 44.55796498523391]
```

**Ejercicio 6** La respuesta del ejercicio 6 es que, se encontro que los minimos valor en x, y son:

```
[0, 2]
```